RecoCTPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc

```diff
@@ -0,0 +1,391 @@
+      (double *) ofs.getXiValues().data(), ofs.getXiValues().size());
+    rpod.s_y_d_vs_xi = ofs.getSplines()[LHCOpticalFunctionsSet::eyd];
+    rpod.s_v_y_vs_xi = ofs.getSplines()[LHCOpticalFunctionsSet::evy];
+    rpod.s_L_y_vs_xi = ofs.getSplines()[LHCOpticalFunctionsSet::eLy];
+
+    // calculate auxiliary data
+    LHCOpticalFunctionsSet::Kinematics k_in = { 0., 0., 0., 0., 0. };
+    LHCOpticalFunctionsSet::Kinematics k_out;
+    rpod.optics->transport(k_in, k_out);
+    rpod.x0 = k_out.x;
+    rpod.y0 = k_out.y;
+
+    // TODO: the fits below could be replaced with simple algebra in order to avoid
+    // the relatively heavy creation of TGraph and fit with TF1
+
+    unique_ptr<TGraph> g_x_d_vs_xi = make_unique<TGraph>(ofs.getXiValues().size(), ofs.getXiValues().data(),
+      ofs.getFcnValues()[LHCOpticalFunctionsSet::exd].data());
+    ff->SetParameters(0., 0.);
+    g_x_d_vs_xi->Fit(ff.get(), "Q");
+    rpod.ch0 = ff->GetParameter(0) - rpod.x0;
+    rpod.ch1 = ff->GetParameter(1);
+
+    unique_ptr<TGraph> g_L_x_vs_xi = make_unique<TGraph>(ofs.getXiValues().size(), ofs.getXiValues().data(),
+      ofs.getFcnValues()[LHCOpticalFunctionsSet::eLx].data());
+    ff->SetParameters(0., 0.);
+    g_L_x_vs_xi->Fit(ff.get(), "Q");
+    rpod.la0 = ff->GetParameter(0);
+    rpod.la1 = ff->GetParameter(1);
+
+    // insert record
+    const CTPPSDetId rpId(p.first);
+    m_rp_optics_.emplace(rpId, std::move(rpod));
+  }
+
+  // update settings
+  initialized_ = true;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void ProtonReconstructionAlgorithm::release()
+{
+  initialized_ = false;
+
+  m_rp_optics_.clear();
+}
+
+//----------------------------------------------------------------------------------------------------
+
+double ProtonReconstructionAlgorithm::ChiSquareCalculator::operator() (const double* parameters) const
+{
+  // extract proton parameters
+  const LHCOpticalFunctionsSet::Kinematics k_in = { 0., parameters[1], parameters[3], parameters[2], parameters[0] };
+
+  // calculate chi^2 by looping over hits
+  double S2 = 0.;
+
+  for (const auto &track : *tracks) {
+    const CTPPSDetId rpId(track->getRPId());
+
+    // transport proton to the RP
+    auto oit = m_rp_optics->find(rpId);
+    LHCOpticalFunctionsSet::Kinematics k_out;
+    oit->second.optics->transport(k_in, k_out);
+
+    // proton position wrt. beam
+    const double& x = k_out.x - oit->second.x0;
+    const double& y = k_out.y - oit->second.y0;
+
+    // make sure that uncertainties are reasonable
+    double x_unc = track->getXUnc();
+    if (x_unc < 1E-3)
+      x_unc = 40E-3;
+
+    double y_unc = track->getYUnc();
+    if (y_unc < 1E-3)
+      y_unc = 40E-3;
+
+    // calculate chi^2 contributions, convert track data mm --> m
+    const double x_diff_norm = (x - track->getX()*1E-3) / (x_unc*1E-3);
+    const double y_diff_norm = (y - track->getY()*1E-3) / (y_unc*1E-3);
```

Comment:
This would be rather a large change - all TOTEM code used mm (or m) and it stayed in CMSSW, too. Furthermore, making the change would break backward compatibility. Are we sure we want to do this? Would it be possible to restrict the change to objects from lite tracks onwards?

Reply:
it's probably OK to keep positions in mm at the local level, given the historical precedent.
My comment was primarily to avoid adding "m" in the mix of units

Reply:
The optical functions now use cm as the length unit.